src/go/mongolib/stats/stats.go

```diff
@@ -0,0 +1,268 @@
+package stats
+
+import (
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter"
+	"github.com/percona/percona-toolkit/src/go/mongolib/proto"
+	"github.com/percona/percona-toolkit/src/go/mongolib/util"
+)
+
+// New creates new instance of stats with given fingerprinter
+func New(fingerprinter fingerprinter.Fingerprinter) *Stats {
+	s := &Stats{
+		fingerprinter: fingerprinter,
+	}
+
+	s.Reset()
+	return s
+}
+
+// Stats is a collection of MongoDB statistics
+type Stats struct {
+	// dependencies
+	fingerprinter fingerprinter.Fingerprinter
+
+	// internal
+	queryInfoAndCounters map[GroupKey]*QueryInfoAndCounters
+}
+
+// Reset clears the collection of statistics
+func (s *Stats) Reset() {
+	s.queryInfoAndCounters = make(map[GroupKey]*QueryInfoAndCounters)
+}
+
+// Add adds proto.SystemProfile to the collection of statistics
+func (s *Stats) Add(doc proto.SystemProfile) error {
+	fp, err := s.fingerprinter.Fingerprint(doc.Query)
+	if err != nil {
+		return fmt.Errorf("cannot get fingerprint: %s", err)
+	}
+	var qiac *QueryInfoAndCounters
+	var ok bool
+
+	key := GroupKey{
+		Operation:   doc.Op,
+		Fingerprint: fp,
+		Namespace:   doc.Ns,
+	}
+	if qiac, ok = s.queryInfoAndCounters[key]; !ok {
```

Comment:
Just a question without analyzing all the flow: don't we need a mutex here? Add is an exported function and it could be called concurrently but Go maps are not thread safe.